### ISSUE TITLE
Improve pystone measurement

### ISSUE THREAD

**Issue by sanderjo** [sabnzbd/sabnzbd#1162]
On my 2018 Core i3 I see a lot of varying pystone values reported by SABnzbd: from 26.696 to 143.348. 

When I run the same python's pystone in standalone and for a longer time, I get higher and more correct values of 165.000 - 170.000:

```
$ /usr/bin/time -f "Real: %e" python -c "from test.pystone import pystones; print int(pystones(15000)[1])"  
167199
Real: 0.13
```

On top of that: I think the duration `1000` (which I introduced some years ago) in
 https://github.com/sabnzbd/sabnzbd/blob/31c0c239f9bc163b968f9d7f19c3b59cd6aa368d/sabnzbd/utils/getperformance.py#L46  is too short on normal CPUs, and certainly on modern fast CPU's.

So I'll create a PR to solve this and get more steady and accurate pystone measurements.

| CPU | Typical Pystone Value |
| --- | --- |
| ARMv8 nanopineo2 | 15.000 |
| Atom N270 (2009) | 25.000 |
| Celeron (2015) | 60.000 |
| Xeon E5335 (2007) | 84.000 | 
| Core i3 (2018) | 170.000 |

**Comment by sanderjo** [comment]
A separate problem: the pystone reporting via the wrench symbol is much lower. @Safihre ... any idea how that can happen? Does the disk speed measurement run at the same time, eating CPU power?

![image](https://user-images.githubusercontent.com/1273502/43370340-b7707ed6-937d-11e8-95e4-29354ff244c3.png)

**Comment by thezoggy** [comment]
Pystone is known to produce varying results due to compiler optimizations and interrupter differences (python flavor).

**Comment by thezoggy** [comment]
Thus running compiled or not.. or cpython vs something else.. your gonna get different results.. looking online people mention upto a 30% delta

**Comment by Safihre** [comment]
> A separate problem: the pystone reporting via the wrench symbol is much lower. @Safihre ... any idea how that can happen? Does the disk speed measurement run at the same time, eating CPU power?

It runs separate before the disk-speed measures ([see here](https://github.com/sabnzbd/sabnzbd/blob/develop/sabnzbd/interface.py#L2434-L2444)).
However, @thezoggy is right: its performance depends on interrupts by other threads. So for example having the interface open sends API-calls all the time, which could overlap with the PyStone, causing it to be interrupted. For perfect measures it should be run outside the GIL.

**Comment by sanderjo** [comment]
An update:

I've created a better pystone()
https://github.com/sabnzbd/sabnzbd/blob/31c0c239f9bc163b968f9d7f19c3b59cd6aa368d/sabnzbd/utils/getperformance.py#L41)
very accurate results when called from the startup procedure in SABnzbd.py. So that's good.

... but calling that pystone() from the GUI still results in lower and varying numbers. I've tried:

- a time.sleep() before and after the call. No improvement
- no disk-calls ... No improvement
- call of 1, 2 or 3 pystone() from other processeses (`import multiprocessing`) ... No improvement

So something is givind low prio to that process. Or the system is really busy?

@Safihre tips?

**Comment by Safihre** [comment]
Try setting the GUI-refresh to 1 minute.

**Comment by sanderjo** [comment]
> Try setting the GUI-refresh to 1 minute.

To be safe, I set it to 30 minutes. Result: 35.000 pystones (instead of 160.000 at startup), with "improved" method.

I'll dig further

**Comment by sanderjo** [comment]
As they say: meten is weten, gissen is missen.

So: in pystone() I create a loop with some different input values:
```
            checklist = [1000, 2000, 5000, 10000, 20000, 50000, 100000, 200000]
            for check in checklist:
                nowval = int(pystones(check)[1])
                print "SJ0:", check, nowval
                if nowval > value:
                    value = nowval
            print "SJ1: best pysthone score is", value
```
Results:

At start of SABnzbd.py:
```
2018-08-02 20:46:52,081::DEBUG::[SABnzbd:1223] Could not determine my IPv6 address
SJ0: 1000 66019
SJ0: 2000 86835
SJ0: 5000 147427
SJ0: 10000 155168
SJ0: 20000 157598
SJ0: 50000 162755
SJ0: 100000 166670
SJ0: 200000 167098
SJ1: best pysthone score is 167098
2018-08-02 20:46:54,480::DEBUG::[SABnzbd:1229] CPU Pystone available performance = 167098
```
pystone(1000) gives a too low value. That's what we do now, and what I solved in my PR-to-be.


From the wrench symbol:
```
SJ0: 1000 26367
SJ0: 2000 33867
SJ0: 5000 99202
SJ0: 10000 109336
SJ0: 20000 133255
SJ0: 50000 152322
SJ0: 100000 125900
SJ0: 200000 136599
SJ1: best pysthone score is 152322
```
Best score at 50.000. 
and
```
SJ0: 1000 23174
SJ0: 2000 33606
SJ0: 5000 105999
SJ0: 10000 127432
SJ0: 20000 126980
SJ0: 50000 148214
SJ0: 100000 118649
SJ0: 200000 123945
SJ1: best pysthone score is 148214
```
Best score at 50.000
and
```
SJ0: 1000 25068
SJ0: 2000 29762
SJ0: 5000 105708
SJ0: 10000 140040
SJ0: 20000 133659
SJ0: 50000 101633
SJ0: 100000 116835
SJ0: 200000 121960
SJ1: best pysthone score is 140040
```
Best score at 10.000

So no need for 100.000 or 200.000 as input. Good; as they take long. Remember: if your CPU's pystone is X, running pystone(X) takes exactly 1 second. So better to stay lower than that, too avoid wasting time.

**Comment by Safihre** [comment]
@sanderjo Tried the new code. Always get a stable ~180.000 via command-line but via SABnzbd between 120.000-180.000, similar to the old code. 
Although now with thew new code it's more often closer to 180.000, so that seems to be an improvement 👍

**Comment by Safihre** [comment]
@sanderjo any development on this? Will you create a PR? 🥇

**Comment by sanderjo** [comment]
I shall create a PR with the code I sent you, and you tested.

OK?

**Comment by Safihre** [comment]
Closed by #1171
